overlord/snapshotstate/snapshotmgr.go

```diff
@@ -0,0 +1,288 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapshotstate
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/overlord/snapshotstate/backend"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+var (
+	osRemove             = os.Remove
+	snapstateCurrentInfo = snapstate.CurrentInfo
+	configGetSnapConfig  = config.GetSnapConfig
+	configSetSnapConfig  = config.SetSnapConfig
+	backendOpen          = backend.Open
+	backendSave          = backend.Save
+	backendRestore       = (*backend.Reader).Restore // TODO: look into using an interface instead
+	backendCheck         = (*backend.Reader).Check
+	backendRevert        = (*backend.RestoreState).Revert // ditto
+	backendCleanup       = (*backend.RestoreState).Cleanup
+)
+
+// SnapshotManager takes snapshots of active snaps
+type SnapshotManager struct{}
+
+// Manager returns a new SnapshotManager
+func Manager(st *state.State, runner *state.TaskRunner) *SnapshotManager {
+	runner.AddHandler("save-snapshot", doSave, doForget)
+	runner.AddHandler("forget-snapshot", doForget, nil)
+	runner.AddHandler("check-snapshot", doCheck, nil)
+	runner.AddHandler("restore-snapshot", doRestore, undoRestore)
+	runner.AddCleanup("restore-snapshot", cleanupRestore)
+
+	manager := &SnapshotManager{}
+	snapstate.AddAffectedSnapsByAttr("snapshot-setup", manager.affectedSnaps)
+
+	return manager
+}
+
+// Ensure is part of the overlord.StateManager interface.
+func (SnapshotManager) Ensure() error { return nil }
+
+func (SnapshotManager) affectedSnaps(t *state.Task) ([]string, error) {
+	if k := t.Kind(); k == "check-snapshot" || k == "forget-snapshot" {
+		// check and forget don't affect snaps
+		// (this could also be written k != save && k != restore, but it's safer this way around)
+		return nil, nil
+	}
+	var snapshot snapshotSetup
```

Comment:
the tests are mocking too much and as a result the bit below is not tested, you probably want at least one test not mocking the conflict change conflict checker, see other managers for examples

Reply:
this is proving to be a lot harder than I expected.

Reply:
shouldn't be

Reply:
ikr

Reply:
something like this:

https://github.com/snapcore/snapd/blob/master/overlord/ifacestate/ifacestate_test.go#L3633

?

Reply:
Done. I think.